AppController/djinn.rb

```diff
@@ -1849,33 +1849,37 @@ def job_start(secret)
         APPS_LOCK.synchronize { versions_loaded_now = @versions_loaded.clone }
 
         # Starts apps that are not running yet but they should.
+        versions_to_load = versions_loaded_now - my_versions_loaded
         if my_node.is_shadow?
           begin
             versions_to_load = ZKInterface.get_versions - versions_loaded_now
           rescue FailedZooKeeperOperationException
             versions_to_load = []
           end
-        else
-          versions_to_load = versions_loaded_now - my_versions_loaded
         end
         versions_to_load.each { |version_key|
           APPS_LOCK.synchronize {
             setup_app_dir(version_key, true)
             setup_appengine_version(version_key)
           }
         }
-        # In addition only shadow kick off the autoscaler.
-        APPS_LOCK.synchronize { scale_deployment if my_node.is_shadow? }
-      end
 
-      # Load balancers and shadow need to check/update applications that have
-      # been undeployed and nginx/haproxy.
-      if my_node.is_shadow? or my_node.is_load_balancer?
-        APPS_LOCK.synchronize {
-          check_stopped_apps
-        }
-      end
-      if my_node.is_load_balancer?
+        # In addition only shadow kick off the autoscaler and manages
+        # running instances.
+        if my_node.is_shadow?
+          missing_nodes = 0
+          @state_change_lock.synchronize {
+            missing_nodes = Integer(@options['min_machines']) - @nodes.length
+          }
+          APPS_LOCK.synchronize {
+            scale_deployment
+            scale_up_instances(missing_nodes) if missing_nodes > 0
+          }
+        end
+
+        # Load balancers need to check/update applications that have been
+        # undeployed and nginx/haproxy.
+        APPS_LOCK.synchronize { check_stopped_apps }
```

Comment:
Doesn't the shadow need to run `check_stopped_apps`?

Reply:
Good question: on the quick answer, shadow is a load_balancer and in the the short term we cannot put it on a non-load-balancer machine. Long term  I would like to have shadow to potentially be on all roles, but it will take a bit work to get there.

Reply:
I think it's important to maintain the mapping we had from role to function. This improves readability and makes future changes much easier. If I were to move all of the functions that the shadow performs to a separate process, I might miss this bit since there's no reference to shadow here.